File: src/AquaHub.T5LampsController/schedule.py
import json
import os
import time
from datetime import datetime, timedelta, time as dt_time
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler


class ScheduleEntry:
    def __init__(self, time, coral_plus, blue_plus):
        self.time = time
        self.coral_plus = coral_plus
        self.blue_plus = blue_plus
        self.parsed_time = self.parse_time(time)

    def parse_time(self, time_str):
        return datetime.strptime(time_str, "%H:%M:%S").time()


class Schedule:
    _demo_mode_expire_time = None
    _current_schedule = []

    def __init__(self):
        self.load_schedule("ReefMoreBlue")
        self.setup_watcher()

    def setup_watcher(self):
        observer = Observer()
        handler = FileSystemEventHandler()
        handler.on_modified = self.on_changed
        observer.schedule(handler, path='.', patterns=['schedule.json'])
        observer.start()

    def on_changed(self, event):
        self.load_schedule("ReefMoreBlue")

    def load_schedule(self, schedule_name):
        with open('schedule.json', 'r') as f:
            schedules = json.load(f)
        self._current_schedule = [ScheduleEntry(**entry) for entry in schedules[schedule_name]]

    def calculate_current_values(self):
        now = datetime.now().time()
        current_blue_plus = None
        current_coral_plus = None

        for i in range(len(self._current_schedule) - 1):
            start = self._current_schedule[i].parsed_time
            end = self._current_schedule[i + 1].parsed_time

            if start <= now <= end:
                if self._current_schedule[i].blue_plus is not None or self._current_schedule[i + 1].blue_plus is not None:
                    current_blue_plus = int(self.interpolate(
                        self._current_schedule[i].blue_plus, self._current_schedule[i + 1].blue_plus, start, end, now))

                if self._current_schedule[i].coral_plus is not None or self._current_schedule[i + 1].coral_plus is not None:
                    current_coral_plus = int(self.interpolate(
                        self._current_schedule[i].coral_plus, self._current_schedule[i + 1].coral_plus, start, end, now))

        return [current_blue_plus or 0, current_coral_plus or 0]

    def interpolate(self, start_value, end_value, start_time, end_time, current_time):
        start_value = start_value or 0
        end_value = end_value or 0

        fraction = ((datetime.combine(datetime.today(), current_time) - datetime.combine(datetime.today(), start_time)).total_seconds() /
                    (datetime.combine(datetime.today(), end_time) - datetime.combine(datetime.today(), start_time)).total_seconds())
        return start_value + (end_value - start_value) * fraction

    def set_demo_mode(self, seconds=None):
        self._demo_mode_expire_time = datetime.now() + timedelta(seconds=seconds or 30)

    def is_demo_mode(self):
        return self._demo_mode_expire_time > datetime.now()

File: src/AquaHub.NConsole/Schedule.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AquaHub.NConsole
{
    public static class Schedule
    {
        private static DateTime? _demoModeExpireTime = null;
        private static List<ScheduleEntry> _currentSchedule;
        private static readonly FileSystemWatcher watcher = new();

        static Schedule()
        {
            watcher.Path = Directory.GetCurrentDirectory();
            watcher.Filter = "schedule.json";
            watcher.NotifyFilter = NotifyFilters.LastWrite;
            watcher.Changed += OnChanged;
            watcher.EnableRaisingEvents = true;

            _currentSchedule = LoadSchedule("ReefMoreBlue");
        }

        private static void OnChanged(object sender, FileSystemEventArgs e)
        {
            _currentSchedule = LoadSchedule("ReefMoreBlue");
        }

        public static Dictionary<string, int> get_current_values()
        {
            return CalculateCurrentValues(_currentSchedule);
        }

        private static List<ScheduleEntry> LoadSchedule(string scheduleName)
        {
            var json = File.ReadAllText("schedule.json");
            var schedules = JsonSerializer.Deserialize<Dictionary<string, List<ScheduleEntry>>>(json);
            foreach (var entry in schedules?[scheduleName]!)
                entry.ParsedTime = TimeSpan.Parse(entry.Time);
            return schedules[scheduleName];
        }

        private static Dictionary<string, int> CalculateCurrentValues(IReadOnlyList<ScheduleEntry> schedule)
        {
            var now = DateTime.Now.TimeOfDay;
            int? currentBluePlus = null;
            int? currentCoralPlus = null;
            int? currentPhotoRed = null;

            for (var i = 0; i < schedule.Count - 1; i++)
            {
                var start = schedule[i].ParsedTime;
                var end = schedule[i + 1].ParsedTime;

                if (start <= now && now <= end)
                {
                    if (schedule[i].BluePlus.HasValue || schedule[i + 1].BluePlus.HasValue)
                        currentBluePlus = (int)Interpolate(schedule[i].BluePlus, schedule[i + 1].BluePlus, start, end, now);

                    if (schedule[i].CoralPlus.HasValue || schedule[i + 1].CoralPlus.HasValue)
                        currentCoralPlus = (int)Interpolate(schedule[i].CoralPlus, schedule[i + 1].CoralPlus, start, end, now);

                    if (schedule[i].PhotoRed.HasValue || schedule[i + 1].PhotoRed.HasValue)
                        currentPhotoRed = (int)Interpolate(schedule[i].PhotoRed, schedule[i + 1].PhotoRed, start, end, now);
                }
            }

            var sched = new Dictionary<string, int>
            {
                { "BluePlus", currentBluePlus ?? 0 },
                { "CoralPlus", currentCoralPlus ?? 0 },
                { "PhotoRed", currentPhotoRed ?? 0 }
            };

            //return new int[] { currentBluePlus ?? 0, currentCoralPlus ?? 0, currentPhotoRed ?? 0 };
            return sched;
        }

        private static double Interpolate(int? startValue, int? endValue, TimeSpan startTime, TimeSpan endTime, TimeSpan currentTime)
        {
            startValue ??= 0;
            endValue ??= 0;

            double fraction = (currentTime - startTime).TotalSeconds / (endTime - startTime).TotalSeconds;
            return startValue.Value + (endValue.Value - startValue.Value) * fraction;
        }

        public static void set_demo_mode(double? seconds = null)
        {
            _demoModeExpireTime = DateTime.Now.AddSeconds(seconds ?? 30);
        }

        public static bool is_demo_mode()
        {
            return _demoModeExpireTime > DateTime.Now;
        }
    }

    public class ScheduleEntry
    {
        [JsonPropertyName("t")]
        public string Time { get; set; }

        public int? CoralPlus { get; set; }

        public int? BluePlus { get; set; }

        public int? PhotoRed { get; set; }

        [JsonIgnore]
        public TimeSpan ParsedTime { get; set; }
    }
}
